File: showcases/SmartOperationsDashboard/smartops-web/src/services/deviceService.ts
import { API_BASE_URL } from '../api/config'

export interface Device {
  id: string
  name: string
  serialNumber: string
  location: string
  status: 'Offline' | 'Online' | 'Idle' | 'Running' | 'Error' | 'Maintenance'
  description?: string
  createdAt: string
  updatedAt: string
  createdBy: string
}

export interface Reading {
  id: string
  deviceId: string
  temperature: number
  humidity: number
  pressure: number
  powerConsumption: number
  operationCount: number
  recordedAt: string
}

export interface DeviceSummary {
  totalDevices: number
  activeDevices: number
  offlineDevices: number
  recentReadings: Reading[]
}

export interface DeviceFilter {
  status?: string
  location?: string
  page?: number
  pageSize?: number
}

export interface DeviceListResponse {
  devices: Device[]
  total: number
  page: number
  pageSize: number
}

class DeviceService {
  private baseUrl = `${API_BASE_URL}/devices`

  async getDevices(filter?: DeviceFilter): Promise<DeviceListResponse> {
    const params = new URLSearchParams()
    if (filter?.status) params.append('status', filter.status)
    if (filter?.location) params.append('location', filter.location)
    if (filter?.page) params.append('page', filter.page.toString())
    if (filter?.pageSize) params.append('pageSize', filter.pageSize.toString())

    const url = params.toString() ? `${this.baseUrl}?${params}` : this.baseUrl
    const response = await fetch(url, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch devices: ${response.statusText}`)
    }

    return response.json()
  }

  async getDevice(id: string): Promise<Device> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch device: ${response.statusText}`)
    }

    return response.json()
  }

  async getSummary(): Promise<DeviceSummary> {
    const response = await fetch(`${API_BASE_URL}/devices/summary`, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch summary: ${response.statusText}`)
    }

    return response.json()
  }

  async createDevice(data: Omit<Device, 'id' | 'createdAt' | 'updatedAt' | 'createdBy'>): Promise<Device> {
    const response = await fetch(this.baseUrl, {
      method: 'POST',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })

    if (!response.ok) {
      throw new Error(`Failed to create device: ${response.statusText}`)
    }

    return response.json()
  }

  async updateDevice(id: string, data: Partial<Omit<Device, 'id' | 'createdAt' | 'updatedAt' | 'createdBy'>>): Promise<Device> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'PUT',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })

    if (!response.ok) {
      throw new Error(`Failed to update device: ${response.statusText}`)
    }

    return response.json()
  }

  async deleteDevice(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'DELETE',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to delete device: ${response.statusText}`)
    }
  }

  async archiveDevice(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/${id}/archive`, {
      method: 'POST',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to archive device: ${response.statusText}`)
    }
  }


  async unarchiveDevice(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/${id}/unarchive`, {
      method: 'POST',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to unarchive device: ${response.statusText}`)
    }
  }

  async getDevicesByUserId(userId: string): Promise<Device[]> {
    const response = await fetch(`${this.baseUrl}/user/${userId}`, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch devices by user ID: ${response.statusText}`)
    }

    return response.json()
  }

  async getDevicesByArchived(userId: string): Promise<Device[]> {
    const response = await fetch(`${this.baseUrl}/user/${userId}/archived`, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch devices by user ID: ${response.statusText}`)
    }

    return response.json()
  }

  // getReadings
  async getReadings(deviceId: string): Promise<Reading[]> {
    const response = await fetch(`${this.baseUrl}/${deviceId}/readings`, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch readings for device ID: ${response.statusText}`)
    }

    return response.json()
  }

  async createReading(deviceId: string, reading: Reading): Promise<Reading> {
    const response = await fetch(`${this.baseUrl}/${deviceId}/readings`, {
      method: 'POST',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(reading),
    })

    if (!response.ok) {
      throw new Error(`Failed to create reading for device ID: ${response.statusText}`)
    }

    return response.json()
  }

  async updateReading(deviceId: string, readingId: string, reading: Reading): Promise<Reading> {
    const response = await fetch(`${this.baseUrl}/${deviceId}/readings/${readingId}`, {
      method: 'PUT',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(reading),
    })

    if (!response.ok) {
      throw new Error(`Failed to update reading for device ID: ${response.statusText}`)
    }

    return response.json()
  }

  async deleteReading(deviceId: string, readingId: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/${deviceId}/readings/${readingId}`, {
      method: 'DELETE',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to delete reading for device ID: ${response.statusText}`)
    }
  }

  async getAllReadings(startDate?: Date, endDate?: Date): Promise<Reading[]> {
    const readingsUrl = `${API_BASE_URL}/readings`
    const params = new URLSearchParams()
    if (startDate) params.append('startDate', startDate.toISOString())
    if (endDate) params.append('endDate', endDate.toISOString())

    const url = params.toString() ? `${readingsUrl}?${params}` : readingsUrl
    const response = await fetch(url, {
      method: 'GET',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
    })

    if (!response.ok) {
      throw new Error(`Failed to fetch readings: ${response.statusText}`)
    }
    const data = await response.json()
    console.log(data.readings)
    return data.readings || []
  }
}

export const deviceService = new DeviceService()

File: showcases/SmartOperationsDashboard/SmartOps.Api/Services/DeviceService.cs
using Microsoft.EntityFrameworkCore;
using SmartOps.Api.Data;
using SmartOps.Api.DTOs;
using SmartOps.Api.Models;

namespace SmartOps.Api.Services;

/// <summary>
/// Device service implementation
/// </summary>
public class DeviceService : IDeviceService
{
    private readonly SmartOpsDbContext _dbContext;

    public DeviceService(SmartOpsDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    public async Task<(List<DeviceDto> devices, int total)> GetDevicesAsync(DeviceFilterDto filter)
    {
        var query = _dbContext.Devices.AsQueryable();

        // Filter by status
        if (!string.IsNullOrEmpty(filter.Status))
        {
            if (Enum.TryParse<DeviceStatus>(filter.Status, true, out var status))
            {
                query = query.Where(d => d.Status == status);
            }
        }

        // Filter by location
        if (!string.IsNullOrEmpty(filter.Location))
        {
            query = query.Where(d => d.Location.Contains(filter.Location));
        }

        var total = await query.CountAsync();

        var devices = await query
            .OrderByDescending(d => d.UpdatedAt)
            .Skip((filter.Page - 1) * filter.PageSize)
            .Take(filter.PageSize)
            .Select(d => new DeviceDto
            {
                Id = d.Id,
                Name = d.Name,
                SerialNumber = d.SerialNumber,
                Location = d.Location,
                Status = d.Status.ToString(),
                Description = d.Description,
                CreatedAt = d.CreatedAt,
                UpdatedAt = d.UpdatedAt,
                CreatedBy = d.CreatedBy,
            })
            .ToListAsync();

        return (devices, total);
    }

    public async Task<DeviceDto?> GetDeviceByIdAsync(Guid id)
    {
        var device = await _dbContext.Devices.FindAsync(id);
        if (device == null)
            return null;

        return new DeviceDto
        {
            Id = device.Id,
            Name = device.Name,
            SerialNumber = device.SerialNumber,
            Location = device.Location,
            Status = device.Status.ToString(),
            Description = device.Description,
            CreatedAt = device.CreatedAt,
            UpdatedAt = device.UpdatedAt,
            CreatedBy = device.CreatedBy,
        };
    }

    public async Task<DeviceDto> CreateDeviceAsync(CreateDeviceDto dto, Guid createdBy)
    {
        // Parse status from DTO, default to Offline if not provided or invalid
        var status = DeviceStatus.Offline;
        if (!string.IsNullOrEmpty(dto.Status))
        {
            if (Enum.TryParse<DeviceStatus>(dto.Status, true, out var parsedStatus))
            {
                status = parsedStatus;
            }
        }

        var device = new Device
        {
            Id = Guid.NewGuid(),
            Name = dto.Name,
            SerialNumber = dto.SerialNumber,
            Location = dto.Location,
            Description = dto.Description,
            Status = status,
            CreatedBy = createdBy,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };

        _dbContext.Devices.Add(device);
        await _dbContext.SaveChangesAsync();

        return new DeviceDto
        {
            Id = device.Id,
            Name = device.Name,
            SerialNumber = device.SerialNumber,
            Location = device.Location,
            Status = device.Status.ToString(),
            Description = device.Description,
            CreatedAt = device.CreatedAt,
            UpdatedAt = device.UpdatedAt,
            CreatedBy = device.CreatedBy,
        };
    }

    public async Task<DeviceDto> UpdateDeviceAsync(Guid id, UpdateDeviceDto dto)
    {
        var device = await _dbContext.Devices.FindAsync(id);
        if (device == null)
            throw new KeyNotFoundException($"Device {id} not found");

        if (!string.IsNullOrEmpty(dto.Name))
            device.Name = dto.Name;

        if (!string.IsNullOrEmpty(dto.Location))
            device.Location = dto.Location;

        if (dto.Description != null)
            device.Description = dto.Description;

        if (!string.IsNullOrEmpty(dto.Status))
        {
            if (Enum.TryParse<DeviceStatus>(dto.Status, true, out var status))
                device.Status = status;
        }

        device.UpdatedAt = DateTime.UtcNow;
        await _dbContext.SaveChangesAsync();

        return new DeviceDto
        {
            Id = device.Id,
            Name = device.Name,
            SerialNumber = device.SerialNumber,
            Location = device.Location,
            Status = device.Status.ToString(),
            Description = device.Description,
            CreatedAt = device.CreatedAt,
            UpdatedAt = device.UpdatedAt,
            CreatedBy = device.CreatedBy,
        };
    }

    public async Task DeleteDeviceAsync(Guid id)
    {
        var device = await _dbContext.Devices.FindAsync(id);
        if (device == null)
            throw new KeyNotFoundException($"Device {id} not found");

        _dbContext.Devices.Remove(device);
        await _dbContext.SaveChangesAsync();
    }

    public async Task<object> GetSummaryAsync()
    {
        var totalDevices = await _dbContext.Devices.CountAsync();
        var activeDevices = await _dbContext
            .Devices.Where(d => d.Status == DeviceStatus.Online || d.Status == DeviceStatus.Running)
            .CountAsync();
        var offlineDevices = await _dbContext
            .Devices.Where(d => d.Status == DeviceStatus.Offline)
            .CountAsync();

        var recentReadings = await _dbContext
            .Readings.OrderByDescending(r => r.RecordedAt)
            .Take(10)
            .Select(r => new
            {
                r.Id,
                r.DeviceId,
                r.Temperature,
                r.Humidity,
                r.Pressure,
                r.PowerConsumption,
                r.OperationCount,
                r.RecordedAt,
            })
            .ToListAsync();

        return new
        {
            totalDevices,
            activeDevices,
            offlineDevices,
            recentReadings,
        };
    }
}
